### PR TITLE
Update samba add-on to use Bashio #735

### LINKED ISSUE #735: Update samba add-on to use Bashio
The add-ons now have [Bashio](https://github.com/hassio-addons/bashio) available, which makes it easier and nicer to work with and is easier to understand. The `samba` add-on needs to be rewritten to use the Bashio functions.

In general, this only requires changes to the add-on [`run.sh`](https://github.com/home-assistant/hassio-addons/blob/master/samba/run.sh) script. A good example of an add-on that has been migrated already, is [`deconz`](https://github.com/home-assistant/hassio-addons/blob/master/deconz/data/run.sh).

Please note, nothing has to be done to make Bashio available, it is already baked-in into the add-on base images.

### PR DESCRIPTION
**Pull request by Zapfmeister**
Still have an error, which im looking for some feedback in order to fix it:

`sed: unmatched '|'`

caused by:
`sed -i "s|%%ALLOW_HOSTS%%|$ALLOW_HOSTS|g" /etc/smb.conf`

closes #735

### CODE REVIEW

**Comment by Zapfmeister** [comment]
closes #735

**Comment by Zapfmeister** [comment]
> Still have an error, which im looking for some feedback in order to fix it:
> 
> `sed: unmatched '|'`
> 
> caused by:
> `sed -i "s|%%ALLOW_HOSTS%%|$ALLOW_HOSTS|g" /etc/smb.conf`

Fixed

File: samba/run.sh [hunk 1]
```diff
@@ -1,38 +1,43 @@
...
 else
-    echo "[INFO] Read hostname: ${NAME}"
+    bashio::log.info "Read hostname: ${NAME}"
 fi
 
+# Workaround to create usable IPRanges from ALLOW_HOSTS variable
+IPRANGES=
+for item in $ALLOW_HOSTS; do
+         IPRANGES="$IPRANGES ${item}"
+     done
```
Comment (frenck) [inline samba/run.sh]: `ALLOW_HOSTS=$(bashio::config "allow_hosts | join(\" \")")` ?

File: samba/run.sh [hunk 2]
```diff
@@ -1,38 +1,43 @@
...
-sed -i "s|%%INTERFACE%%|$INTERFACE|g" /etc/smb.conf
-sed -i "s|%%ALLOW_HOSTS%%|$ALLOW_HOSTS|g" /etc/smb.conf
-sed -i "s|%%USERNAME%%|$USERNAME|g" /etc/smb.conf
+sed -i "s|%%WORKGROUP%%|${WORKGROUP}|g" /etc/smb.conf
+sed -i "s|%%NAME%%|${NAME}|g" /etc/smb.conf
+sed -i "s|%%INTERFACE%%|${INTERFACE}|g" /etc/smb.conf
+sed -i "s|%%USERNAME%%|${USERNAME}|g" /etc/smb.conf
+sed -i "s#%%ALLOW_HOSTS%%#${IPRANGES}#g" /etc/smb.conf
 
+cat /etc/smb.conf
```
Comment (frenck) [inline samba/run.sh]: Forgotten debug statement?

File: samba/run.sh [hunk 3]
```diff
@@ -1,38 +1,43 @@
...
+INTERFACE=$(bashio::config 'interface')
+ALLOW_HOSTS=$(bashio::config 'allow_hosts')
+USERNAME=$(bashio::config 'username')
+PASSWORD=$(bashio::config 'password')
 
 WAIT_PIDS=()
 NAME=
 
 # Check Login data
 if [ -z "${USERNAME}" ] || [ -z "${PASSWORD}" ]; then
```
Comment (frenck) [inline samba/run.sh]: Example:
`if ! bashio::config.has_value 'username'; then`

File: samba/run.sh [hunk 4]
```diff
@@ -1,38 +1,43 @@
...
-    echo "[ERROR] No valid login data inside options!"
+    bashio::log.error "No valid login data inside options!"
     exit 1
 fi
 
 # Read hostname from API
 if ! NAME="$(curl -s -f -H "X-Hassio-Key: ${HASSIO_TOKEN}" http://hassio/info | jq --raw-output '.data.hostname')"; then
```
Comment (frenck) [inline samba/run.sh]:
```
NAME=$(bashio::info.hostname)
if bashio::var.is_empty "${NAME}"; then
   bashio::log.warn "Can't read hostname, using default."
fi
bashio::log.info "Hostname: ${NAME}"
```

File: samba/run.sh [hunk 5]
```diff
@@ -1,37 +1,35 @@
...
+PASSWORD=$(bashio::config 'password')
 
 WAIT_PIDS=()
 NAME=
 
 # Check Login data
-if [ -z "${USERNAME}" ] || [ -z "${PASSWORD}" ]; then
-    echo "[ERROR] No valid login data inside options!"
+if ! bashio::config.has_value 'username' || ! bashio::config.has_value 'password'; then
+    bashio::log.error "No valid login data inside options!"
```
Comment (frenck) [inline samba/run.sh]:
```suggestion
    bashio::exit.nok "No valid login data inside options!"
```

File: samba/run.sh [hunk 6]
```diff
@@ -1,37 +1,35 @@
...
     exit 1
```
Comment (frenck) [inline samba/run.sh]:
```suggestion
```

**Comment by frenck** [comment]
Nice! Last changes and I will set up a final test locally to ensure it all works.
After that, we can start merging it in!

**Pull request merged**